### PR TITLE
Rate-limit password-based login attempts

### PR DESCRIPTION
**Pull request by sandhose**
This is a first step for #2541.

It sets up a `Limiter` utility, which will hold all the rate-limiting logic.
As a first candidate, I rate-limited password-based logins.

There are two levels of rate-limiting:

 - per-IP address, which has a low limit (3 per minute)
 - per-user, which has a higher limit (1800 per hour)

I'm hesitant about the latter, but this is to try avoid distributed password-guessing attacks.
You'd need about 600 IPs to effectively lock out a user, which should be fine?

### CODE REVIEW

File: crates/handlers/src/rate_limit.rs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+// limitations under the License.
+
+use std::{net::IpAddr, sync::Arc};
+
+use governor::{clock::QuantaClock, state::keyed::DashMapStateStore, Quota, RateLimiter};
+use mas_data_model::User;
+use nonzero_ext::nonzero;
+use ulid::Ulid;
+
+const PASSWORD_CHECK_FOR_REQUESTER_QUOTA: Quota = Quota::per_minute(nonzero!(3u32));
```
Comment (reivilibre) [inline crates/handlers/src/rate_limit.rs:22]: Not really sure either way but 3/min seems a bit harsh, just thinking about e.g. the company's shared office which has a shared IPv4.

File: crates/handlers/src/rate_limit.rs [hunk 2]
```diff
@@ -0,0 +1,183 @@
...
+#[derive(Debug)]
+struct LimiterInner {
+    password_check_for_requester: KeyedRateLimiter<RequesterFingerprint>,
+    password_check_for_user: KeyedRateLimiter<Ulid>,
+}
+
+impl Default for LimiterInner {
+    fn default() -> Self {
+        Self {
+            password_check_for_requester: RateLimiter::keyed(PASSWORD_CHECK_FOR_REQUESTER_QUOTA),
```
Comment (reivilibre) [inline crates/handlers/src/rate_limit.rs:79]: can't remember for sure but I think you want to call `retain_recent` on these ratelimiters occasionally to perform housekeeping.

**Review by reivilibre — APPROVED** [review]
overall this seems right to me!

File: crates/handlers/src/compat/login.rs [hunk 3]
```diff
@@ -650,6 +670,57 @@ mod tests {
         assert_eq!(body, old_body);
     }
 
+    /// Test that password logins are rate limited.
+    #[sqlx::test(migrator = "mas_storage_pg::MIGRATOR")]
```
[the remark this answers is not in the document]
Comment (sandhose) [reply crates/handlers/src/compat/login.rs:674]: It's nice, it creates a temporary database and run the migrator on it, so each test gets an isolated database

File: crates/handlers/src/rate_limit.rs (hunk 1 above)
Comment (sandhose) [reply crates/handlers/src/rate_limit.rs:22]: I'll do config and tweak those numbers in another PR

File: crates/handlers/src/rate_limit.rs (hunk 2 above)
Comment (sandhose) [reply crates/handlers/src/rate_limit.rs:79]: Good call, I did that in 5b1c9b45

**Comment by sandhose** [comment]
@reivilibre if you could glimpse at 5b1c9b45 and retick, that would be great

**Review by reivilibre — APPROVED** [review]

**Pull request merged**